evalml/automl/utils.py

```diff
@@ -193,7 +193,7 @@ def get_best_sampler_for_data(X, y, sampler_method, sampler_balanced_ratio):
             import_or_raise(
                 "imblearn.over_sampling", error_msg="imbalanced-learn is not installed"
             )
-            cat_cols = X.ww.select("Categorical").columns
+            cat_cols = X.ww.select(exclude="numeric").columns
```

Comment:
I think this will cause a problem if users pass in non-categorical and non-numeric features that get converted to numeric by components in the pipeline. 

For example, this counts text as categoricals. But the text would get turned into doubles by the text featurizer. So they wouldn't be categoricals by the time it got to the sampler. That would cause a stacktrace because SmoteNC expects categoricals to be present, right?

```python
from evalml.pipelines import BinaryClassificationPipeline
from evalml.automl.utils import get_best_sampler_for_data
import pandas as pd
import woodwork as ww
import pytest

X = pd.DataFrame({"a": ["Hello this is my text"] * 50 + ["Hello this is my bad text"] * 50,
                  "b": list(range(100)),
                  "c": pd.date_range("2020-01-01", periods=100)})
y = pd.Series([1] * 90 + [0] * 10)
X.ww.init(logical_types={"a": "NaturalLanguage", "c": "Datetime"})

sampler = get_best_sampler_for_data(X, y, sampler_method="Oversampler",
                                    sampler_balanced_ratio=0.5)

pipeline = BinaryClassificationPipeline({"Dt": ["DateTime Featurization Component", "X", "y"],
                                         "Text": ["Text Featurization Component", "Dt.x", "y"],
                                         "Oversampler": [sampler, "Text.x", "y"],
                                         "RF": ["Random Forest Classifier", "Oversampler.x", "Oversampler.y"]
                                        })

with pytest.raises(IndexError):
    pipeline.fit(X, y)
```

![image](https://user-images.githubusercontent.com/41651716/128562856-d74209e5-34a3-4f36-b8d6-861447f2d2ee.png)

Reply:
Discussed with @freddyaboulton, we agreed the base issue is that what's considered numeric/categorical/neither when we select the proper sampler is different than what's numeric/categorical when the data has made it to the sampler in the pipeline. 

Right now, `boolean` columns are the only non-numeric, non-categorical columns that are considered categorical by SMOTE. Therefore for the purposes of this PR, the fix is just to select for both categorical and boolean columns when determining which SMOTE Oversampler to use. Freddy will file an issue to track future work to make oversampler selection more scalable, since that is out of scope for this PR.

Reply:
Issue filed here!  https://github.com/alteryx/evalml/issues/2605